subsys/mgmt/mcumgr/smp_bt.c

```diff
@@ -41,6 +56,101 @@ static struct bt_uuid_128 smp_bt_svc_uuid = BT_UUID_INIT_128(
 static struct bt_uuid_128 smp_bt_chr_uuid = BT_UUID_INIT_128(
 	BT_UUID_128_ENCODE(0xda2e7828, 0xfbce, 0x4e01, 0xae9e, 0x261174997c48));
 
+/* Helper function that allocates conn_param_data for a conn. */
+static struct conn_param_data *alloc_conn_param_data(struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(conn_data); i++) {
+		if (conn_data[i].conn == NULL) {
+			conn_data[i].conn = conn;
+			return &conn_data[i];
+		}
+	}
+
+	/* Conn data must exists. */
+	__ASSERT_NO_MSG(false);
+	return NULL;
+}
+
+/* Helper function that returns conn_param_data associated with a conn. */
+static struct conn_param_data *get_conn_param_data(const struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(conn_data); i++) {
+		if (conn_data[i].conn == conn) {
+			return &conn_data[i];
+		}
+	}
+
+	/* Conn data must exists. */
+	__ASSERT_NO_MSG(false);
+	return NULL;
+}
+
+/* Work handler function for restoring the default latency for the connection. */
+static void restore_default_latency(struct k_work *work)
+{
+	struct bt_le_conn_param params;
+	struct conn_param_data *cpd;
+	struct bt_conn_info info;
+	int ret = 0;
+
+	cpd = CONTAINER_OF(work, struct conn_param_data, dwork);
+
+	if (cpd->latency_state & CONN_LOW_LATENCY_REQUESTED) {
+		(void)k_work_reschedule(&cpd->dwork, K_MSEC(RESTORE_TIME));
+		cpd->latency_state &= ~CONN_LOW_LATENCY_REQUESTED;
+	} else {
+		ret = bt_conn_get_info(cpd->conn, &info);
+		if (ret != 0) {
+			return;
+		}
+
+		if (info.le.latency >= cpd->default_latency) {
+			return;
+		}
+
+		params.interval_min = info.le.interval;
+		params.interval_max = info.le.interval;
+		params.latency = cpd->default_latency;
+		params.timeout = info.le.timeout;
+
+		ret = bt_conn_le_param_update(cpd->conn, &params);
+
+		if (ret != 0 && (ret != -EALREADY)) {
+			(void)k_work_reschedule(&cpd->dwork, K_MSEC(RESTORE_RETRY_TIME));
+		}
+	}
+}
+
+/* Enables low latency connection for a given conn. */
+static void enable_low_latency(struct bt_conn *conn)
+{
+	struct bt_le_conn_param params;
+	struct conn_param_data *cpd;
+	struct bt_conn_info info;
+	int ret = 0;
+
+	cpd = get_conn_param_data(conn);
+	if (cpd->latency_state == 0) {
```

Comment:
This would not work properly if connection slave latency is set to 0 and requested flag was cleared by the work (in that case flag needs to be set, but we should not send the request)

Reply:
I forgot to clear the flags in the work, when the latency is restored. In case you are describing the work would exit at first if:
```
		if (ret != 0 || info.le.latency == 0) {
			return;
		}

Reply:
ACtually i do it in `le_param_updated` callback so i think there is no need to add this.

Reply:
If i understand you correctly the answer is: It would work as there is ENABLED flag that is not cleared by the work.